### PR TITLE
mgr: perf schema fns/change notification and Prometheus plugin

### PR DESCRIPTION
**Pull request by dmick**

### CODE REVIEW

**Comment by dmick** [comment]
Possible discussion points:

1. disabled by default, must set port
2. does not currently set timestamps; easy to do but perhaps inadvisable
3. does minimal translation on stat names; there may be better mappings
4. does not attempt to label stats with any dimension other than 'daemon.id', but could easily do so

**Comment by dmick** [comment]
retest this please

File: src/mgr/PyModules.cc [hunk 1]
```diff
@@ -687,6 +687,62 @@ PyObject* PyModules::get_counter_python(
...
+    if (daemon_state.exists(key)) {
+      states[key] = daemon_state.get(key);
+    }
+  }
+
+  PyFormatter f;
+  f.open_object_section("perf_schema");
+
+  // FIXME: this is unsafe, I need to either be inside DaemonStateIndex's
+  // lock or put a lock on individual DaemonStates
```
Comment (liewegas) [inline src/mgr/PyModules.cc]: Hmm, should fix this before merge.  John, any preference how to approach this?  Maybe a DaemonState dump method that takes the DaemonStateCollection and Formatter as an arg and locks appropriately?

Comment (dmick) [inline src/mgr/PyModules.cc]: Yeah, I confess I just boilerplated and did not understand the issue

**Comment by liewegas** [comment]
1. I like the pattern of having a default port and binding to :: by default so that a simple 'ceph mgr module enable prometheus' is sufficient to turn it on.

**Comment by dmick** [comment]
...and now that I'm using cherrypy I could do that, I suppose.  The Prometheus client library didn't accept that (needed 0.0.0.0).

**Comment by l-mb** [comment]
Very interesting! Is the goal to make this feature equivalent (or better) to ceph_exporter, or is it meant to be used alongside it?

Will it perhaps even expose the same metrics as a drop-in replacement? (Or should we take this discussion to ceph-devel?)

**Comment by dmick** [comment]
l-mb: the goal is to get something in for enhancement :D  I do know about ceph-exporter, and realized I'd need to read the code, which means reading Go, to really understand what it advertises and how.  It would probably make more sense to try to be as compatible as possible in terms of metric names, types, and labels, I agree.  I think the intent is to make a "low-barrier" collection point for small-to-medium clusters.  But yes, happy to discuss on ceph-devel too

**Comment by dmick** [comment]
repushed with default listen addr set to '::'

File: src/mgr/PyModules.cc (hunk 1 above)
Comment (liewegas) [inline src/mgr/PyModules.cc]: I have a separate PR that adds a per-DaemonState lock, so we can ignore that issue here.

**Comment by dmick** [comment]
Hmm, maybe should pick a port like these:  https://github.com/prometheus/prometheus/wiki/Default-port-allocations  I suspect we don't want to share 9128 in case in future we have both mon/osd and mgr on the same host and want to use per-daemon instrumentation as in ceph-exporter

**Comment by liewegas** [comment]
:+1:

**Comment by dmick** [comment]
Added 9283 there, updated default port here

**Comment by jan--f** [comment]
@dmick I was going to suggest to use port 9128 actually. If this code exports what the ceph_exporter produces (and more) then the two are compatible from the scrapers perspective, i.e. the mgr prometheus plugin could be a drop-in replacement for the ceph_exporter.

**Comment by jan--f** [comment]
[See here](http://paste.opensuse.org/33014291) for an example ceph_exporter output.

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-07-20_05:37:32-rados-wip-sage-testing2-distro-basic-smithi/ is fine.

just need to add the doc before merge!

**Comment by dmick** [comment]
@jan--f: but if you want both on one host, that's a problem.  (i.e. if the ceph-mgr version does only cluster-wide stats, and the ceph_exporter version does the daemons.)

**Comment by dmick** [comment]
@liewegas Docs?!?  what are those???

um....er, yeah.  <digs to see where and what>

**Comment by dmick** [comment]
Added docs.  There are also recommendations for Prometheus metric names (see https://prometheus.io/docs/practices/naming/), which I think ceph_exporter uses, that would require a translation table of some kind here.   I don't know how important it is.  Perhaps this is a "put this out there and refine it going forward" type question.

**Comment by jan--f** [comment]
@dmick I would think the mgr prometheus code will eventually export much more then the ceph_exporter can. The ceph_exporter can only export what it finds in the return of ```ceph health``` and in the osdmap and pgmap. But as you say this is probably also a "put this out there and refine it going forward" type question.
One minor thing: Its common practice for prometheus exporters to expose their metrics under ```<host>:<port>/metrics```. If I understand your code correctly, the plugin exposes its metrics under ```/```. https://prometheus.io/docs/instrumenting/writing_exporters/#landing-page

**Comment by liewegas** [comment]
I think there is a lot of cleanup we can do in the metric names.  I like that doc because it's prescriptive about it, but adopting probably means going in and renaming a bunch of our internal perfcounters (vs having some ugly remapping table at this layer).  So, longer term effort...

**Comment by dmick** [comment]
@jan--f: ceph_exporter doesn't use 'daemon perf dump'?  I'd just assumed it would.  I'm sure that's an easy-ish mod.

I mentioned in the doc, but it actually just ignores path, so /metrics works, as does /; it's the same technique the Python Prometheus client library used, so I didn't stress.  Easy enough to add path handling if/when we need it (@jcsp mentioned that federation might be a thing, so /federate might be a future).  

@liewegas: yeah, mapping in this app was what I was thinking.

**Pull request merged**